### DISCUSSION TITLE
Getvalues() return old form values

### DISCUSSION THREAD

**Discussion by Temkit** [react-hook-form/react-hook-form#3813] [Q&A]
I have a form component, and a Hoc Component with all the the react hook form routines.

When I route from the form component to the same form component changing only the data from the query 
Ex : list/key/id1 to list/key/id2 and run getvalues() return the id1 data even though the the form is filled with the id2 data ....

Someone have any idea 💡 :)

**Comment by bluebill1049** [comment]
provide a reproduable CSB?

**Reply by Temkit** [reply]
thanks for your reply, due to very big/complex project I can't really reproduce...
but i installed the RHF debug, when I go to the second route, the UI changes with the new Data but RHF debug keeps the old data....

FIRST FORM BEFORE ROUTING
<img width="1064" alt="Capture d’écran 2021-01-01 à 16 42 15" src="https://user-images.githubusercontent.com/13275723/103441819-cfa93680-4c50-11eb-92c6-51db94d2e283.png">

SECOND FORM AFTER ROUTING / UI DARA CHANGES BUT THE RHF DEBUG NOT ( when I route there from a non form component everything is fine .... )
<img width="1062" alt="Capture d’écran 2021-01-01 à 16 42 56" src="https://user-images.githubusercontent.com/13275723/103441823-dcc62580-4c50-11eb-8390-1a220c029704.png">

**Reply by bluebill1049** [reply]
looks like you are using a single useForm all the entire application, i think you should break into multiple useForms

**Reply by Temkit** [reply]
yes ! this is my HOC ,  my Idea is to have a DRY form component ! what can you advice to have a new useform each time spec changes, i mean one can easly think of useeffect, but Hook on Hooks .... ! 

knowing i have more than a hundred form (it will grow to more), i thought of a json spec form, which i render through my HOC.
i mean create a hundred form class with each one a unique useform is not good idea !
```

import React, { useContext, useMemo, useEffect, useState } from "react";
import AppContext from "context";
import { useMutation } from "@apollo/client";
import { useForm } from "react-hook-form";
import { getOperationName } from "apollo-boost";

import { DevTool } from "@hookform/devtools";
import { useRouter } from "next/router";

const withManualFormHOC = (WrappedComponent) => (props) => {
  const { setAppSpec } = useContext(AppContext);
  const router = useRouter();

  let [Create, { error: createError, called }] = useMutation(
    props.spec?.create
  );
  let [Save, { error: saveError }] = useMutation(props?.spec.save);

  const { register, setValue, getValues } = useForm({
    defaultValues: props.item,
  });

  const submit = async () => {
    const values = getValues();

    console.log(values);

    let value_to_store = {};
    for (const key in values) {
      if (values[key] !== null && values[key] !== undefined) {
        props.spec.fields.forEach((tab) => {
          tab.attributes.forEach((element) => {
            if (element.name === key) {
              switch (element.type) {
                case "number":
                  if (element.step.includes(".")) {
                    value_to_store[key] = parseFloat(values[key]);
                  } else {
                    value_to_store[key] = parseInt(values[key]);
                  }
                  break;
                case "checkbox":
                  value_to_store[key] = values[key] ? true : false;
                  break;
                default:
                  value_to_store[key] = values[key];
                  break;
              }
            }
          });
        });
      }
    }

    if (!value_to_store.id) {
      const variables = {
        data: value_to_store,
      };

      await Create({
        variables,
        refetchQueries: [getOperationName(props.spec.refresh)],
      });
    } else {
      const id = values.id;

      delete values.id;
      const variables = {
        variables: { id, data: values },
      };

      if (props.spec.refresh) {
        variables.refetchQueries = [getOperationName(props.spec.refresh)];
      }

      await Save(variables);
    }

    if (createError || saveError) {
      console.log(saveError);
      setAppSpec((state) => {
        state.toastit = {
          type: "danger",
          text: createError.message + saveError.message,
        };
      });
    } else {
      setAppSpec((state) => {
        state.toastit = {
          type: "success",
          text: "élèment enrgistré avec succés",
        };
      });
    }

    if (props.spec.back) {
      router.back();
    }
  };

  return (
    <>
      <WrappedComponent
        {...props}
        register={register}
        setValue={setValue}
        submit={submit}
      />
    </>
  );
};

export default withManualFormHOC;

```

**Reply by bluebill1049** [reply]
then you should consider `unregister` input which not part of the form data.

**Reply by Temkit** [reply]
hmmm. thanks. in fact when I changed my navigation function form next.js route push to window.location.href = "" it's working!
can you help me understand that !
